### PR TITLE
Accept Variable as argument for NumpyArray

### LINKED ISSUE #546: can't pass a Variable to NumpyArray
pyccel crashes when we try to pass a variable to it (C and fortran)

```python
import numpy as np

a = np.array([1,2,3])
b = np.array(a)
```
the error message is :
```bash/fortran/c
  File "/home/bauom/.app/pyccel/pyccel/ast/numpyext.py", line 137, in __new__
    raise TypeError('Uknown type of  %s.' % type(arg))
TypeError: Uknown type of  <class 'pyccel.ast.core.Variable'>.
```

### PR DESCRIPTION
**Pull request by MenoIy**
Fix #546:
- accept variable as an argument in NumpyArray;
- update NumpyArray C printer;
- add tests.

### CODE REVIEW

File: pyccel/ast/numpyext.py [hunk 1]
```diff
@@ -133,11 +133,12 @@ class NumpyArray(NumpyNewArray):
...
 
-        if not isinstance(arg, (PythonTuple, PythonList)):
+        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
             raise TypeError('Uknown type of  %s.' % type(arg))
 
         # TODO: treat inhomogenous lists and tuples when they have mixed ordering
-        if not arg.is_homogeneous:
+        if isinstance(arg, (PythonTuple, PythonList)) and not arg.is_homogeneous or \
+            isinstance(arg, Variable) and not arg.is_ndarray and not arg.is_stack_array:
             raise TypeError('we only accept a homogeneous list or tuple ')
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:142]: Don't forget to update the error message

File: pyccel/ast/numpyext.py [hunk 2]
```diff
@@ -133,11 +133,12 @@ class NumpyArray(NumpyNewArray):
     def __init__(self, arg, dtype=None, order='C'):
         NumpyNewArray.__init__(self)
 
-        if not isinstance(arg, (PythonTuple, PythonList)):
+        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
             raise TypeError('Uknown type of  %s.' % type(arg))
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:137]: There is a typo here `Uknown`->`Unknown`. Would you mind fixing it?

File: pyccel/ast/numpyext.py [hunk 3]
```diff
@@ -133,12 +133,13 @@ class NumpyArray(NumpyNewArray):
...
-            raise TypeError('Uknown type of  %s.' % type(arg))
+        if not isinstance(arg, (PythonTuple, PythonList, Variable)):
+            raise TypeError('Unknown type of  %s.' % type(arg))
 
         # TODO: treat inhomogenous lists and tuples when they have mixed ordering
-        if not arg.is_homogeneous:
-            raise TypeError('we only accept a homogeneous list or tuple ')
+        if isinstance(arg, (PythonTuple, PythonList)) and not arg.is_homogeneous or \
+            isinstance(arg, Variable) and not arg.is_ndarray and not arg.is_stack_array:
+            raise TypeError('we only accept a homogeneous arguments ')
```
Comment (EmilyBourne) [inline pyccel/ast/numpyext.py:142]: `we only accept homogeneous arguments`

File: pyccel/codegen/printing/ccode.py [hunk 4]
```diff
@@ -1120,15 +1120,21 @@ def _print_Assign(self, expr):
             if rhs.rank == 0:
                 raise NotImplementedError(expr.lhs + "=" + expr.rhs)
             dummy_array_name, _ = create_incremented_string(self._parser.used_names, prefix = 'array_dummy')
-            dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+            declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+            dtype = self.find_in_ndarray_type_registry(format(rhs.dtype), rhs.precision)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:1124]: What does the call to `format` do here?

Comment (MenoIy) [reply pyccel/codegen/printing/ccode.py:1124]: my bad it should be self._print()

**Review by EmilyBourne — APPROVED** [review]

**Review by mostafaMamoni — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**